File: q0238/q0238.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int n = nums.length; 

        int[] prefixProducts = new int[n];
        prefixProducts[0] = nums[0];
        for (int i = 1; i < n; i++) {
            prefixProducts[i] = prefixProducts[i-1] * nums[i];
        }

        int[] suffixProducts = new int[n];
        suffixProducts[n-1] = nums[n-1];
        for (int i = n-2; i >= 0; i--) {
            suffixProducts[i] = suffixProducts[i+1] * nums[i];
        }

        int[] res = new int[n];
        for (int i = 0; i < n; i++) {
            int prefix = i > 0 ? prefixProducts[i-1] : 1; 
            int suffix = i < n-1 ? suffixProducts[i+1] : 1; 
            res[i] = prefix * suffix;
        }
        return res;
    }
}

File: q0238/q0238.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        prefixProducts = [0 for _ in nums]
        prefixProducts[0] = nums[0]
        for i in range(1, len(nums)): 
            prefixProducts[i] = prefixProducts[i-1] * nums[i]
        
        suffixProducts = [0 for _ in nums]
        suffixProducts[-1] = nums[-1]
        for i in range(len(nums)-2, -1, -1):
            suffixProducts[i] = suffixProducts[i+1] * nums[i]
        
        res = [0 for _ in nums]
        for i in range(len(nums)):
            prefix = prefixProducts[i-1] if i > 0 else 1
            suffix = suffixProducts[i+1] if i < len(nums)-1 else 1
            res[i] = prefix * suffix
        return res 